### PR TITLE
Feature/add last modified field for all entities

### PR DESCRIPTION
**Pull request by katybaulch**
# Description

Added created/last_modified columns for family, collection, event, and document entities, including updating alembic migration scripts to automatically generate the DDL for the triggers and procedures used to execute this functionality.

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

## How Has This Been Tested?

Application level tests currently in admin-backend, database level tests to come in backend repo.

## Reviewer Checklist

- [ ] The PR represents a single feature (small driveby fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: tests/core/ingestion/test_family.py [hunk 1]
```diff
@@ -211,7 +215,6 @@ def test_family_document_from_row__updates_status(test_db: Session, patch_curren
         2000, 1, 1, 0, 0, 0, 0, tzinfo=timezone.utc
     )
     assert db_family_doc.last_modified is not None
-    assert db_family_doc.created != db_family_doc.last_modified
```
Comment (THOR300) [inline tests/core/ingestion/test_family.py:214]: For context I believe @diversemix likes this way of building up the tests into stronger assertions from a loose initial assertion, but not an issue!

File: docker-compose.dev.yml [hunk 2]
```diff
@@ -1,9 +1,10 @@
-version: '3.7'
+version: "3.7"
 services:
-
   backend:
     environment:
       PYTHONDONTWRITEBYTECODE: 0
+    env_file:
+      - .env
```
Comment (THOR300) [inline docker-compose.dev.yml:7]: How come this is now needed?

File: app/db/models/law_policy/family.py [hunk 3]
```diff
@@ -208,7 +217,7 @@ class FamilyDocument(Base):
     )
     last_modified = sa.Column(
         sa.DateTime(timezone=True),
-        default=sa.func.now(),
+        server_default=sa.func.now(),
```
Comment (THOR300) [inline app/db/models/law_policy/family.py:220]: Out of interest why the move to server_default?

File: app/core/config.py [hunk 4]
```diff
@@ -6,14 +6,16 @@
...
+if PUBLIC_APP_URL:
+    PUBLIC_APP_URL = PUBLIC_APP_URL.rstrip("/")
 API_V1_STR = "/api/v1"
 
 # OpenSearch Config
-OPENSEARCH_URL = os.environ["OPENSEARCH_URL"]
-OPENSEARCH_USERNAME = os.environ["OPENSEARCH_USER"]
-OPENSEARCH_PASSWORD = os.environ["OPENSEARCH_PASSWORD"]
-OPENSEARCH_INDEX_PREFIX = os.environ["OPENSEARCH_INDEX_PREFIX"]
+OPENSEARCH_URL = os.getenv("OPENSEARCH_URL", "")
```
Comment (THOR300) [inline app/core/config.py:15]: Why this change, when would we want empty values for these as default?

File: alembic/versions/0021_added_created_and_last_modified_columns_.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+"""
+import sqlalchemy as sa
+from alembic_utils.pg_function import PGFunction
+from alembic_utils.pg_trigger import PGTrigger
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "0021"
```
Comment (olaughter) [inline alembic/versions/0021_added_created_and_last_modified_columns_.py:16]: One thought, that I'm not sure about but wanted to mention. That sometimes it would be tempting to separate out larger migrations like this into per-table changes. This has the downside of extra clutter as now there are more files, but makes individual ones smaller so easier to read and also means a smoother rollback, as if one of them had a problem you wouldnt have to roll all of them back. I don't have strong views on this so no pressure. What do you think?

**Review by olaughter — COMMENTED** [review]
Nice work! Had a few minor thoughts/questions below, shout if any of them don't make sense!

**Review by THOR300 — COMMENTED** [review]
Looks great, just some clarification questions from me! Also I'm no sql alchemy wizard but the tests look very reliable!  

Do you have a link to the linear ticket or something, would love to read around about the context of family events etc.!

File: alembic/env.py [hunk 6]
```diff
@@ -131,7 +230,26 @@ def run_migrations_online():
...
 
 
-register_entities([last_modified_procedure, last_modified_trigger])
+register_entities(
+    # [
+    #     last_modified_procedure,
+    #     doc_last_modified_trigger,
+    #     event_last_modified_trigger,
+    #     collection_last_modified_trigger,
+    # ]
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply alembic/env.py:239]: Great spot!

File: app/db/models/law_policy/family.py (hunk 3 above)
Comment (katybaulch) [reply app/db/models/law_policy/family.py:220]: This is some sqlalchemy [wizardry](https://docs.sqlalchemy.org/en/20/core/defaults.html) :mage_woman: Basically, when you use server_default, the default value (in our case, the current timestamp) will be put into the DDL for the table and therefore inserted into the database if no value is provided. Default on the other hand is used when the value that's going to be inserted into that column is calculated by a function/procedure/trigger. Originally, I thought about using the triggers I've defined to update the values on INSERT and UPDATE, but realised it wasn't necessary to use a trigger for INSERT as I could simply define the default value in the DDL. The logic for when the UPDATE trigger is called though is a heck of a lot more complicated, so using default values isn't the best thing to do.

File: app/core/config.py (hunk 4 above)
Comment (katybaulch) [reply app/core/config.py:15]: Apologies, this config file snuck into the PR. Will remove.

File: app/core/config.py [hunk 7]
```diff
@@ -6,14 +6,16 @@
...
+OPENSEARCH_USERNAME = os.getenv("OPENSEARCH_USER", "")
+OPENSEARCH_PASSWORD = os.getenv("OPENSEARCH_PASSWORD", "")
+OPENSEARCH_INDEX_PREFIX = os.getenv("OPENSEARCH_INDEX_PREFIX", "")
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply app/core/config.py:18]: Apologies, this config file snuck into the PR. Will remove.

File: tests/core/conftest.py [hunk 8]
```diff
@@ -6,32 +6,32 @@
...
 
 @pytest.fixture()
 def patch_current_time():
     return patch_time
 
 
 @contextmanager
-def patch_time(time_to_freeze, tick=True):
+def patch_time(time_to_freeze, entity, tick=False):
     with freeze_time(time_to_freeze, tick=tick) as frozen_time:
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply tests/core/conftest.py:17]: Eyyy I didn't realise, will remove this :)

File: alembic/versions/0021_added_created_and_last_modified_columns_.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+    )
+    op.replace_entity(public_family_document_update_last_modified)  # type: ignore
+
+    public_family_event_update_last_modified = PGTrigger(
+        schema="public",
+        signature="update_last_modified",
+        on_entity="public.family_event",
+        is_constraint=False,
+        definition="""
+    BEFORE UPDATE ON public.family_event
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply alembic/versions/0021_added_created_and_last_modified_columns_.py:139]: Ah, it might seem like it but actually no! The inserts are taken care of by default DDL values. Some things we soft delete (which to the DB is an update statement) and some things we hard delete. We cannot update the last modified time of a family event if the record is deleted from the family_event table, so we instead update the last_modified on family when this happens. It confused the hell out of me for most of this week trying to get the expected behaviour to work! :confused:

File: alembic/versions/0021_added_created_and_last_modified_columns_.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    )
+    op.create_entity(public_family_update_last_modified)  # type: ignore
+
+    public_family_document_update_family_last_modified = PGTrigger(
+        schema="public",
+        signature="update_family_last_modified",
+        on_entity="public.family_document",
+        is_constraint=False,
+        definition="""
+    BEFORE INSERT OR UPDATE OR DELETE ON public.family_document
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply alembic/versions/0021_added_created_and_last_modified_columns_.py:163]: Following on from my reply to your other comment, when we insert a new document into the family document table, we want to update the last_modified time on the family. Same if we update that same document or remove it from the family. This is separate from updating the last_modified time on the document itself.

File: alembic/versions/0021_added_created_and_last_modified_columns_.py (hunk 5 above)
Comment (katybaulch) [reply alembic/versions/0021_added_created_and_last_modified_columns_.py:16]: I've gone back through and separated things out a little more. It's still a bit janky, but considerably better!

File: alembic/versions/0020_added_created_and_last_modified_columns_.py [hunk 11]
```diff
@@ -32,7 +32,12 @@ def upgrade():
     )
     op.add_column(
         "family_document",
-        sa.Column("last_modified", sa.DateTime(timezone=True), nullable=False),
+        sa.Column(
+            "last_modified",
+            sa.DateTime(timezone=True),
+            server_default=sa.text("now()"),
```
[the remark this answers is not in the document]
Comment (katybaulch) [reply alembic/versions/0020_added_created_and_last_modified_columns_.py]: I completely agree that this is the right way to do it -Alembic autogenerate updated this itself. I've had a look on staging and prod and both are using alembic schema version 0019, so we *should* be fine this time. I have gone back in this evening and messed around with using alter_column to add it in 0020 and also tried dropping and readding the column and both of these methods mean that we end up with different created and last_modified times when doing a schema update (when I think I'd expect the data to remain the same). I would need to drop both the created and last_modified columns in 0020 and then recreate them to maintain data integrity. I am happy to do that if needs be, but I think we've gotten lucky this time. Let me know what you think!

File: docker-compose.dev.yml (hunk 2 above)
Comment (katybaulch) [reply docker-compose.dev.yml:7]: Reverted :smile:

File: alembic/versions/0021_added_created_and_last_modified_columns_.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (olaughter) [reply alembic/versions/0021_added_created_and_last_modified_columns_.py]: Ohhh interesting

File: alembic/versions/0021_added_created_and_last_modified_columns_.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (olaughter) [reply alembic/versions/0021_added_created_and_last_modified_columns_.py]: Nice, that makes sense 👍

File: alembic/versions/0020_added_created_and_last_modified_columns_.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (olaughter) [reply alembic/versions/0020_added_created_and_last_modified_columns_.py:38]: Thanks for checking! In that case id expect it to be fine

**Review by olaughter — APPROVED** [review]
lgtm
